### DISCUSSION TITLE
How to best await an async function execution in CyPress?

### DISCUSSION THREAD

**Discussion by mesqueeb** [cypress-io/cypress#15478] [Questions and Help]
My App uses Firebase Authentication.

I have a function attached to the `window` for CyPress, that authenticates a user and waits until the app's user data has been and loaded into my Vue app.

This is the code I use to await this function in my CyPress `before` hook:
```ts
  new Cypress.Promise((resolve: any, reject: any) => {
    cy.window().then((w: any) => {
      w.authConnect
        .signinAndLoadData('cypress-test-user@cypress.io', 'some-password')
        .then(resolve)
        .catch(reject)
    })
  })
```

Is this OK or is there a better way?

PS:
I couldn't find anything on this in the "Real World App" because it only authenticates users either through the UI or through a network call.

**Comment by Kanwar27** [comment]
Probably not an answer but more of a comment:

I personally see nothing wrong with using a Promise , however my personnel preference is adding a [cypress custom command](https://docs.cypress.io/api/cypress-api/custom-commands.html) and then use it with . operator. Cypress by default makes sure that chainnable events are fired in order when they are called with . operator so one less thing for me to worry about.

**Reply by mesqueeb** [reply]
@Kanwar27 would you be able to rewrite my code a a custom command as per an answer? That would be so awesome and appreciated!!

**Reply by Kanwar27** [reply]
Well not 100% sure if this would work but give this a try

```ts
Cypress.Commands.add("signinAndLoadData", () => {
	cy
		.window()
		.then
		(
			(win) => 
			{
				win
					.authConnect
					.signinAndLoadData('cypress-test-user@cypress.io', 'some-password')
			}
		)
})
```

```ts
In your index.d.ts write 

.
.
signinAndLoadData(): Chainable<any>
.
.
```

Finally just use it like any other cypress command within your test file

cy
	.signinAndLoadData()
	...successor cypress commands...
	
Assuming you have not set any custom setting to disable cypress from throwing errors when one arises, then you don't need explicit catch/throw as Cypress would automatically fail test if an error arises. Hope it helps!

**Reply by mesqueeb** [reply]
@Kanwar27 can you add
- above your code: ```ts
- below your code: ```

**Reply by mesqueeb** [reply]
@Kanwar27 thanks a lot!!!
